### PR TITLE
fix issue 20011, 17828 - crash or accept write operation on members of manifest constant structs

### PR DESCRIPTION
**Pull request by ghost**
This fixes a regression introduced by [this fix](https://github.com/dlang/dmd/pull/9505) as well as a related bug.

### CODE REVIEW

**Review by thewilsonator — APPROVED** [review]

**Comment by Basile-z** [comment]
Considering that next  minor release is about to be published, is this still good for stable ?

**Comment by thewilsonator** [comment]
Please target stable.

**Comment by Basile-z** [comment]
:crossed_fingers: no conflicts :crossed_fingers:

**Comment by thewilsonator** [comment]
No conflicts but you've got appveyor there, how old is the parent commit for that branch?

**Comment by Basile-z** [comment]
It must be an Appveyor issue because my local stable branch is up to date (9ddedc4f9f2a64f1077f7f3a73a3aa49eccb3989)

**Comment by Basile-z** [comment]
I will force push otherwise automerge won't happen.

**Comment by kinke** [comment]
Auto-merge 13 mins after opening the PR? - I think this only patches over existing cracks.

**Comment by Basile-z** [comment]
The Lvalue logic is either botched or misused. I had to do this that way because the combination of the two bugs were driving me crazy but honestly this is not to bad: The error check is called in every overridden visit for an assign expression so although the Lvalue logic is not used this patch is clean. It's a leaf, easy to remove, it doesn't impose himself in the middle of something more complex.

**Comment by kinke** [comment]
> The Lvalue logic is either botched or misused.

Yes, something is off since #9505, and that should be analyzed and fixed properly IMO, incl. fixing this:
```D
void main()
{
    const x = 0;
    const p = &x;
    (cast() x) = 5; // wrongly const-folded to 0 = 5
    assert(*p == 5); // fails, still 0
}
```

I guess that as soon as the root problem is understood, we don't need the special cases handled here (we didn't before #9505, did we?).

**Comment by Basile-z** [comment]
Yes this fix is a bit special. There was this 2017 issue I was trying to understand and when minimizing by hand I found the smaller test case that was accepted and then only I realized there was a REG too. Now with this patch all the cases are handled. 

My opinion on the rrot of the issue is that dotted expressions on enum were never really checked, only simple values (using STC.manifest) are.

Anyway I don't think this gonna make it for 2.087. PRs on stable branch are blocked by the project tester.
So no hurry. I chalenge other contributord to find a better fix. But fix the two issues too.

**Comment by wilzbach** [comment]
> Considering that next minor release is about to be published, is this still good for stable ?

Please be a bit careful when pushing to stable in those two days between the last release candidate and the actual release.

**Comment by thewilsonator** [comment]
> PRs on stable branch are blocked by the project tester.

built kite fails for unrelated reasons and is not required.

File: src/dmd/expressionsem.d [hunk 1]
```diff
@@ -11830,3 +11838,31 @@ bool verifyHookExist(const ref Loc loc, ref Scope sc, Identifier id, string desc
...
+/**
+ * Prevent write attempts on members of manifest constants or initializers typed as struct.
+ *
+ * Params:
+ *      exp = An expression that's attempted to be written.
+ *            Must be the LHS of an `AssignExp`, `BinAssignExp` or `CatAssignExp`.
+ * Returns:
+ *      `false` if no attempt detected and an `true` otherwise, after issuing a message.
+ */
+private bool errorIfEnumStructMemberWritten(Expression exp)
```
Comment (WalterBright) [inline src/dmd/expressionsem.d]: In the rest of dmd, `check` is used instead of `error`.

File: src/dmd/expressionsem.d [hunk 2]
```diff
@@ -11830,3 +11838,31 @@ bool verifyHookExist(const ref Loc loc, ref Scope sc, Identifier id, string desc
...
+{
+    // e1.var = ...
+    // e1.ident = ...
+    Expression e1;
+    if (exp.op == TOK.dotVariable)
+        e1 = exp.isDotVarExp().e1;
+    else if (exp.op == TOK.dotIdentifier)
+        e1 = exp.isDotIdExp().e1;
+
```
Comment (WalterBright) [inline src/dmd/expressionsem.d]:
```
else
    return false;
```

File: src/dmd/expressionsem.d [hunk 3]
```diff
@@ -11830,3 +11838,31 @@ bool verifyHookExist(const ref Loc loc, ref Scope sc, Identifier id, string desc
...
+    // enum SomeStruct ss = { ... }
+    // also true for attempt on init: SomeStruct.init.member = ...
+    if (!e1 || e1.op != TOK.structLiteral)
```
Comment (WalterBright) [inline src/dmd/expressionsem.d]: the `!e1` doesn't really belong here.

**Review by WalterBright — CHANGES_REQUESTED** [review]
see comment

File: src/dmd/expressionsem.d [hunk 4]
```diff
@@ -11830,3 +11838,31 @@ bool verifyHookExist(const ref Loc loc, ref Scope sc, Identifier id, string desc
...
 }
+
+/**
+ * Prevent write attempts on members of manifest constants or initializers typed as struct.
+ *
+ * Params:
+ *      exp = An expression that's attempted to be written.
+ *            Must be the LHS of an `AssignExp`, `BinAssignExp` or `CatAssignExp`.
+ * Returns:
+ *      `false` if no attempt detected and an `true` otherwise, after issuing a message.
```
Comment (Basile-z) [inline src/dmd/expressionsem.d]: ~~an~~

**Comment by wilzbach** [comment]
> built kite fails for unrelated reasons and is not required.

Come on. It is not unknown - it lacks a PR from druntime where the mangling was changed for Mecca. That one needs to be cherry-picked to stable.

**Review by RazvanN7 — APPROVED** [review]

File: src/dmd/expressionsem.d [hunk 5]
```diff
@@ -11830,3 +11838,31 @@ bool verifyHookExist(const ref Loc loc, ref Scope sc, Identifier id, string desc
     error(loc, "`%s.%s` not found. The current runtime does not support %.*s, or the runtime is corrupt.", module_.toChars(), id.toChars(), cast(int)description.length, description.ptr);
     return false;
 }
+
+/**
+ * Prevent write attempts on members of manifest constants or initializers typed as struct.
+ *
+ * Params:
+ *      exp = An expression that's attempted to be written.
+ *            Must be the LHS of an `AssignExp`, `BinAssignExp` or `CatAssignExp`.
```
Comment (Basile-z) [inline src/dmd/expressionsem.d]: `ref` and `out` parameters wrongly accept the dot expr too

Comment (Basile-z) [reply src/dmd/expressionsem.d]: and non constant pointer to members should also be prevented

**Comment by Basile-z** [comment]
@WalterBright, can you see if the changes requested are applied correctly ?

**Comment by Basile-z** [comment]
As seen in libasync through the project tester, fixed const ref function parameters which can accept init members as the intention is not to modify.

**Review by WalterBright — APPROVED** [review]

**Comment by anass-O** [comment]
@kinke did you ever revert this? Its causing more bugs and it seems to have introduced (partially) rvalues Bindable to `const ref`.

Also this is the cause of this regression.

Issues.dlang.org/show_bug.cgi?id=20608

**Comment by kinke** [comment]
It would be reverted in #10124, but that's blocked because of the Ocean lib, which needs a new release (waiting for months now).

**Pull request merged**